token delegation

notJoon opened the pull request:
# Description

Introduces the _**Delegated Token**_ implementation, which allows users to deposit tokens, delegate voting rights, and withdraw tokens.

## Changes

- Added new file `delegated_token.go` with the following key functions:
  - `New`: Creates a new `DelegatedToken` instance
  - `Deposit`: Deposit the entire approved amount of tokens
  - `DepositAmount`: Deposit a specific amount of tokens
  - `Withdraw`: Withdraw the entire staked balance
  - `WithdrawAmount`: Withdraw a specific amount of tokens
  - `Delegate`: Delegate voting rights to another address
  - `GetStaker`: Get the address of the associated staker contract
  - `GetDelegatedTo`: Get the address to which an account has delegated
  - `BalanceOf`: Get the balance of delegated tokens for an account

notJoon committed:
token delegation

notJoon committed:
fix deposit

notJoon committed:
withdraw

notJoon committed:
lint

notJoon committed:
update doc

notJoon committed:
apply lint

r3v4s commented on delegation/delegated.gno:
```diff
@@ -0,0 +1,254 @@
+var gDelegatedToken *DelegatedToken
+
+// Init initializes the DelegatedToken contract
+func Init(staker std.Address, token grc20.Token, name string, symbol string) {
```
Can it be `init()` for automatic initialization

notJoon replied:
Renamed it `New` instead of `Init` because I wanted to allows control the `DelegatedToken` instance during runtime. https://github.com/gnoswap-labs/governance/pull/9/commits/3cfbe186ad4276c17998428cac818403e595d56d

r3v4s commented on delegation/delegated.gno:
```diff
@@ -0,0 +1,254 @@
+	if currDelegatee != to {
+		gDelegatedToken.delegatedTo[owner] = to
+
+		// TODO: move delegates ?
```
move to where? for what??

notJoon replied:
I used to write it for notes, so it's not that important. removed https://github.com/gnoswap-labs/governance/pull/9/commits/3cfbe186ad4276c17998428cac818403e595d56d

r3v4s commented on delegation/delegated.gno:
```diff
@@ -0,0 +1,254 @@
+	token := gDelegatedToken.token
+
+	// get the approved amount
+	allowance := token.Allowance(caller, std.GetOrigPkgAddr())
```
Don't use `std.GetOrigPkgAddr()`, but `std.CurrentRealm().Addr()`

notJoon replied:
https://github.com/gnoswap-labs/governance/pull/9/commits/3cfbe186ad4276c17998428cac818403e595d56d

r3v4s reviewed (changes requested):
Test cases run good, left few comments

notJoon committed:
fixup

r3v4s commented on delegation/delegated.gno:
```diff
@@ -0,0 +1,252 @@
+
+// WithdrawAmount withdraws specified amount of token
+// from the contract for the caller
+func WithdrawAmount(amount *u256.Uint) {
```
Seems like user has to call this function directly.
If it is, make amount parameter to string type then convert to u256

notJoon replied:
https://github.com/gnoswap-labs/governance/pull/9/commits/681e41d9529d86d8ee8adacbe9d59a148474fbec

r3v4s commented on delegation/delegated_test.gno:
```diff
@@ -0,0 +1,386 @@
+///////////////////////////////////////////////////////////////////////////////////////
+
+func TestGetStaker(t *testing.T) {
+	mockStaker := testutils.TestAddress("mockStaker")
```
Use testutils to generate valid address

notJoon replied:
https://github.com/gnoswap-labs/governance/pull/9/commits/ea81900449d5679339843915a4ccd666908041ec

r3v4s commented on delegation/delegated.gno:
```diff
@@ -0,0 +1,252 @@
+	name        string
+
+	// symbol is the symbol of the delegated token.
+	symbol      string
```
Seems like this struct includes metadata from grc20 spec. In this case please include decimals too.

+ Do we need getter for this or not?? I'm not sure

notJoon replied:
both added https://github.com/gnoswap-labs/governance/pull/9/commits/ea81900449d5679339843915a4ccd666908041ec

r3v4s commented on delegation/delegated.gno:
```diff
@@ -0,0 +1,252 @@
+	token := gDelegatedToken.token
+
+	// get the approved amount
+	allowance := token.Allowance(caller, std.CurrentRealm().Addr())
```
`std.CurrentRealm().Addr()` vs `std.GetOrigPkgAddr()`

notJoon replied:
fixed to use std.GetOrigPkgAddr()

r3v4s commented on delegation/delegated.gno:
```diff
@@ -0,0 +1,252 @@
+		panic("no allowance")
+	}
+
+	err := token.TransferFrom(caller, std.GetOrigPkgAddr(), allowance)
```
`std.CurrentRealm().Addr()` vs `std.GetOrigPkgAddr()`

notJoon replied:
fixed to use `std.GetOrigPkgAddr()`[](url)

r3v4s commented on delegation/delegated.gno:
```diff
@@ -0,0 +1,252 @@
+	caller := std.GetOrigCaller()
+	balance := BalanceOf(caller)
+
+	if balance.Cmp(amount) < 0 {
```
```suggestion
	if balance.Cmp(amount) <= 0 {
```

no need to withdraw `0` amount

r3v4s replied:
```suggestion
	if balance.Cmp(amount) == -1 {
```

more straight forward
https://github.com/gnolang/gno/blob/1f58ee46fbf76f2696c06299610c9bd3483e9b10/examples/gno.land/p/demo/uint256/cmp.gno#L11-L15

r3v4s reviewed (changes requested):
Few things to change

dongwon8247 commented on delegation/delegated.gno:
```diff
@@ -0,0 +1,252 @@
+	)
+}
+
+// Withdraw withdraws the entire staked balance from the contract for the caller
```
Can we make it to be able to withdraw partially?

notJoon replied:
`WithdrawAmount` does that, but I'll modify it so that it can be handled by a single `Withdraw` function.

Also, I'll do the same with the `Deposit` function.

notJoon replied:
https://github.com/gnoswap-labs/governance/pull/9/commits/681e41d9529d86d8ee8adacbe9d59a148474fbec

dongwon8247 commented on delegation/delegated.gno:
```diff
@@ -0,0 +1,252 @@
+	}
+}
+
+// Deposit transfers the approved amount of the staked token to this contract
```
Can we make it to be able to deposit partially?

notJoon replied:
https://github.com/gnoswap-labs/governance/pull/9/commits/2d18511bdc312b0fb1b26f3ba69d0f69e9eb4591

notJoon committed:
Update delegation/delegated.gno

Co-authored-by: Blake <104744707+r3v4s@users.noreply.github.com>

notJoon committed:
use testutil

notJoon committed:
partial withdraw

notJoon committed:
partial deposit

notJoon committed:
make params as primitives

r3v4s commented:
Requesting one more thing to test.

user => some other realm => delegation contract
- if something like above call stack happens, check if `std.GetOrigPkgAddr()` does return exact value as what you intended.

notJoon committed:
feat: add different realm interactions

notJoon commented:
> Requesting one more thing to test.
> 
> user => some other realm => delegation contract
> 
>     * if something like above call stack happens, check if `std.GetOrigPkgAddr()` does return exact value as what you intended.

I added a test, But not sure if it's right.

https://github.com/gnoswap-labs/governance/pull/9/commits/d7313db6e334fbb5902973fb4925029fda83519e

notJoon committed:
Merge branch 'main' into delegation-token

r3v4s commented:
> I added a test, But not sure if it's right.

It isn't right. To test such scenario you need to mock stack frames
- std.NewUserRealm
- std.NewCodeRealm
- std.TestSetRealm

notJoon committed:
Merge branch 'main' into delegation-token

notJoon committed:
fix test

notJoon committed:
update README

notJoon commented:
> > I added a test, But not sure if it's right.
> 
> It isn't right. To test such scenario you need to mock stack frames
> 
>     * std.NewUserRealm
> 
>     * std.NewCodeRealm
> 
>     * std.TestSetRealm

fixed to use mock methods https://github.com/gnoswap-labs/governance/pull/9/commits/fdaa4f617a4ecf9779471bfe5424ca26da4cde90

notJoon committed:
update delegation flow chart
